### PR TITLE
perf(468): Add metric counters for committed transactions and rollbacks

### LINKED ISSUE #468: Record metric for number of transactions
Label metric with transaction type, database id, and reducer id.

### PR DESCRIPTION
**Pull request by joshua-spacetime**
Closes #468.

# Description of Changes

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

**Review by cloutiertyler — APPROVED** [review]
I'll approve this, but I wonder if it would be a better design to put the `database_id` in the `Datastore` or the trait, and to put the transaction type and transaction info inside the Tx itself. Not sure if that's better tbh though, but food for thought.

Also why use the `database_id` rather than the database's `Address`? Please note that `database_id` should never be surfaced to SpacetimeDB users. It's purely internal. `Address` is the identifier for a database for users.

**Review by cloutiertyler — CHANGES_REQUESTED** [review]
Please replace `database_id` with `address` in the exposed metrics.

File: crates/core/src/db/datastore/traits.rs [hunk 1]
```diff
@@ -378,14 +378,20 @@ pub trait Tx {
...
     fn begin_tx(&self) -> Self::TxId;
-    fn release_tx(&self, tx: Self::TxId);
+    fn release_tx(&self, ctx: &ExecutionContext, tx: Self::TxId);
 }
 
 pub trait MutTx {
     type MutTxId;
 
     fn begin_mut_tx(&self) -> Self::MutTxId;
+    fn rollback_mut_tx_with_context(&self, ctx: &ExecutionContext, tx: Self::MutTxId);
```
Comment (kulakowski) [inline crates/core/src/db/datastore/traits.rs:388]: Should this just stay `rollback_mut_tx`, and the test-only `rollback_mut_tx` get some worse name? Lots of things are going to plump around an ExecutionContext to the point that it doesn't seem exceptional or worth calling out.

File: crates/core/src/db/datastore/traits.rs [hunk 2]
```diff
@@ -378,14 +378,20 @@ pub trait Tx {
     type TxId;
 
     fn begin_tx(&self) -> Self::TxId;
-    fn release_tx(&self, tx: Self::TxId);
+    fn release_tx(&self, ctx: &ExecutionContext, tx: Self::TxId);
```
Comment (kulakowski) [inline crates/core/src/db/datastore/traits.rs:381]: eg it is not `release_tx_with_context`

**Review by kulakowski — APPROVED** [review]
Just a note about naming.

File: crates/core/src/db/datastore/traits.rs (hunk 1 above)
Comment (kulakowski) [reply crates/core/src/db/datastore/traits.rs:388]: Yeah. Like I said, I'm happy to stamp another PR that changes em later if that's easier.

**Review by cloutiertyler — APPROVED** [review]
LGTM

**Pull request merged**